src/core/execute.c

```diff
@@ -2581,9 +2581,18 @@ static int exec_child(
                         [EXEC_DIRECTORY_CONF] = EXIT_CONF_DIRECTORY,
                 };
                 ExecDirectoryType j;
+                uid_t _uid = uid;
+                gid_t _gid = gid;
+
+                if (MANAGER_IS_SYSTEM(unit->manager)) {
+                        if (uid == UID_INVALID)
+                                _uid = 0;
+                        if (gid == GID_INVALID)
+                                _gid = 0;
+                }
```

Comment:
please move this into setup_exec_directory(), I think that'd be prettier.

Also, please use uid_is_valid() for checking whether a uid is valid, it's a bit more careful